### PR TITLE
[Feature] update Scissor Test

### PR DESCRIPTION
**Pull request by Mee-gu**
update Scissor Test

### CODE REVIEW

File: cocos/renderer/backend/metal/CommandBufferMTL.h [hunk 1]
```diff
@@ -45,6 +50,13 @@ class CommandBufferMTL : public CommandBuffer
     RenderPipelineMTL* _renderPipelineMTL = nullptr;
     BindGroup* _bindGroup = nullptr;
     unsigned int _renderTargetHeight = 0;
+    unsigned int _renderTargetWidth = 0;
+    float _defaultScissorX = 0;
+    float _defaultScissorY = 0;
+    float _preScissorRect[4] = {0.0, 0.0, 0.0, 0.0};
+    MTLScissorRect _mtlScissorRect;
+    bool _isScissorEnabled = false;
+    MTLViewport viewport;
```
Comment (minggo) [inline cocos/renderer/backend/metal/CommandBufferMTL.h]: viewport -> _mtlViewport

File: cocos/renderer/backend/metal/CommandBufferMTL.mm [hunk 2]
```diff
@@ -178,7 +179,7 @@ MTLCullMode toMTLCullMode(CullMode mode)
 
 void CommandBufferMTL::setViewport(int x, int y, unsigned int w, unsigned int h)
 {
-    MTLViewport viewport;
+//    MTLViewport viewport;
```
Comment (minggo) [inline cocos/renderer/backend/metal/CommandBufferMTL.mm]: can remove it

File: cocos/renderer/CCRenderer.h [hunk 3]
```diff
@@ -234,6 +234,11 @@ class CC_DLL Renderer
 
     /** returns whether or not a rectangle is visible or not */
     bool checkVisibility(const Mat4& transform, const Size& size);
+    
+    bool isScissorEnable();
```
Comment (minggo) [inline cocos/renderer/CCRenderer.h]: bool isScissorEnable() -> bool isScissorEnable() const

File: cocos/renderer/CCRenderer.h [hunk 4]
```diff
@@ -234,6 +234,11 @@ class CC_DLL Renderer
...
+    void enableScissor(bool enabled);
+    void getScissorRect(float rect[4]);
```
Comment (minggo) [inline cocos/renderer/CCRenderer.h]: it is better to add a type `ScissorRect` in `ccTypes.h`

File: cocos/platform/CCGLView.cpp [hunk 5]
```diff
@@ -269,14 +269,13 @@ bool GLView::isScissorEnabled()
 
 Rect GLView::getScissorRect() const
 {
-    GLfloat params[4];
     auto renderer = Director::getInstance()->getRenderer();
-    renderer->getScissorRect(params);
+    ScissorRect rect = renderer->getScissorRect();
```
Comment (minggo) [inline cocos/platform/CCGLView.cpp]: please use reference instead

File: cocos/renderer/CCRenderer.h [hunk 6]
```diff
@@ -235,9 +241,9 @@ class CC_DLL Renderer
     /** returns whether or not a rectangle is visible or not */
     bool checkVisibility(const Mat4& transform, const Size& size);
     
-    bool isScissorEnable();
+    bool isScissorEnable() const;
     void enableScissor(bool enabled);
-    void getScissorRect(float rect[4]);
+    ScissorRect getScissorRect() const;
```
Comment (minggo) [inline cocos/renderer/CCRenderer.h]: i think return const reference is better

File: cocos/renderer/CCRenderer.h [hunk 7]
```diff
@@ -142,6 +142,12 @@ Whenever possible prefer to use `TrianglesCommand` objects since the renderer wi
 class CC_DLL Renderer
 {
 public:
+    struct ScissorState
+    {
+        ScissorRect rect;
+        bool isEnabled = false;
+    };
```
Comment (minggo) [inline cocos/renderer/CCRenderer.h]: is it need to be public?

File: cocos/renderer/CCRenderer.h [hunk 8]
```diff
@@ -235,9 +241,9 @@ class CC_DLL Renderer
     /** returns whether or not a rectangle is visible or not */
     bool checkVisibility(const Mat4& transform, const Size& size);
     
-    bool isScissorEnable();
+    bool isScissorEnable() const;
     void enableScissor(bool enabled);
```
Comment (minggo) [inline cocos/renderer/CCRenderer.h]: please use the same function style as depth test:
enableScissor -> setScissorTest
isScissorEnable -> getScissorTest

**Comment by Mee-gu** [comment]
update review

**Comment by Mee-gu** [comment]
update review

**Pull request merged**